dbt/adapters/databricks/python_models/python_submissions.py

```diff
@@ -1,70 +1,175 @@
+        self.tracker = tracker
+        self.cluster_id = cluster_id
+
+    @override
+    def submit(self, compiled_code: str) -> None:
+        context_id = self.api_client.command_contexts.create(self.cluster_id)
+        command_exec: Optional[CommandExecution] = None
+        try:
+            command_exec = self.api_client.commands.execute(
+                self.cluster_id, context_id, compiled_code
+            )
+
+            self.tracker.insert_command(command_exec)
+            # poll until job finish
+            self.api_client.commands.poll_for_completion(command_exec)
+
+        finally:
+            if command_exec:
+                self.tracker.remove_command(command_exec)
+            self.api_client.command_contexts.destroy(self.cluster_id, context_id)
+
+
+class PythonNotebookUploader:
+    """Uploads a compiled Python model as a notebook to the Databricks workspace."""
+
+    def __init__(
+        self, api_client: DatabricksApiClient, catalog: str, schema: str, identifier: str
+    ) -> None:
+        self.api_client = api_client
+        self.catalog = catalog
+        self.schema = schema
+        self.identifier = identifier
+
+    @staticmethod
+    def create(
+        api_client: DatabricksApiClient, parsed_model: ParsedPythonModel
+    ) -> "PythonNotebookUploader":
+        return PythonNotebookUploader(
+            api_client,
+            parsed_model.catalog,
+            parsed_model.schema_,
+            parsed_model.identifier,
         )
 
-        self.job_grants: Dict[str, List[Dict[str, Any]]] = self.workflow_spec.pop("grants", {})
+    def upload(self, compiled_code: str) -> str:
+        """Upload the compiled code to the Databricks workspace."""
+        workdir = self.api_client.workspace.create_python_model_dir(self.catalog, self.schema)
+        file_path = f"{workdir}{self.identifier}"
+        self.api_client.workspace.upload_notebook(file_path, compiled_code)
+        return file_path
 
-    def get_timeout(self) -> int:
-        timeout = self.parsed_model["config"].get("timeout", DEFAULT_TIMEOUT)
-        if timeout <= 0:
-            raise ValueError("Timeout must be a positive integer")
-        return timeout
 
-    def check_credentials(self) -> None:
-        self.credentials.validate_creds()
+@dataclass(frozen=True)
+class PythonJobDetails:
+    """Details required to submit a Python job run to Databricks."""
+
+    run_name: str
+    job_spec: Dict[str, Any]
+    additional_job_config: Dict[str, Any]
+
+
+class PythonPermissionBuilder:
+    """Class for building access control list for Python jobs."""
 
-    def _update_with_acls(self, cluster_dict: dict) -> dict:
-        acl = self.parsed_model["config"].get("access_control_list", None)
-        if acl:
-            cluster_dict.update({"access_control_list": acl})
-        return cluster_dict
+    def __init__(
+        self,
+        api_client: DatabricksApiClient,
+        job_grants: Dict[str, List[Dict[str, Any]]],
```

Comment:
If it is builder, should `job_grants` and `acls` be passed to build method rather than constructor

Reply:
Arguably yes.  In this case the builder is one time use, so functionally it doesn't matter, it just moves the process of extracting the job grants into the create method.  However, stylistically, I can see your point.